libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Trace/Tests.hs

```diff
@@ -0,0 +1,244 @@
+      feeCoin = txb ^. feeTxBodyL
+  inputsAction proof (txb ^. inputsTxBodyL)
+  outputsAction proof txb (fmap (TxOutF proof) (toList (txb ^. outputsTxBodyL)))
+  feesAction feeCoin
+  expectedLedgerState <- getTarget (ledgerStateT proof)
+  env1 <- getEnv
+
+  -- Now compute the env we compute using the STS
+  case applySTSByProof proof (TRC (lenv, ledgerstate, tx)) of
+    Right ledgerState' ->
+      pure (ledgerState' === {- `ediffEq` -} expectedLedgerState)
+    Left errs ->
+      pure
+        ( counterexample
+            ("\napplySTS fails\n" ++ unlines (map show errs) ++ "\nTx =\n" ++ (show (pcTx proof tx)))
+            ( whenFail
+                (putStrLn ("\napplySTS fails\n" ++ unlines (map show errs)) >> (goRepl proof env1 ""))
+                (True === False)
+            )
+        )
+
+main1 :: IO ()
+main1 = do
+  hSetEncoding stdout utf8
+  defaultMain $ testProperty "TraceMain" (withMaxSuccess 50 (toGen genAndRunSimpleTx))
+
+-- ==============================================================
+-- Code to make Traces
+
+-- | Iterate a function 'make' to make a trace of length 'n'. Each call to 'make' gets the
+--   most recent value of the Env internal to TraceM. The function 'make' is
+--   supposed to compute 'a', and (possibly) update the Env internal to TraceM.
+makeTrace :: Int -> TraceM era a -> TraceM era [(Env era, a)]
+makeTrace 0 _ = pure []
+makeTrace n make = do
+  env0 <- getEnv
+  a <- make
+  xs <- makeTrace (n - 1) make
+  pure ((env0, a) : xs)
+
+data Trip era a = Trip !(Env era) !(Env era) !a
+
+beforeAfterTrace :: Int -> (Int -> TraceM era a) -> TraceM era [Trip era a]
+beforeAfterTrace 0 _ = pure []
+beforeAfterTrace !n make = do
+  !beforeEnv <- getEnv
+  !a <- make n
+  !afterEnv <- getEnv
+  xs <- beforeAfterTrace (n - 1) make
+  let !ans = Trip beforeEnv afterEnv a : xs
+  pure ans
+
+-- =================================================================
+-- Show that each step in a trace computes the right LedgerState
+
+runOne ::
+  ( Environment (EraRule "LEDGER" era) ~ LedgerEnv era
+  , State (EraRule "LEDGER" era) ~ LedgerState era
+  , Signal (EraRule "LEDGER" era) ~ Tx era
+  , Show (PredicateFailure (EraRule "LEDGER" era))
+  , Reflect era
+  ) =>
+  Proof era ->
+  TxIx ->
+  Trip era (Signal (EraRule "LEDGER" era)) ->
+  Typed Property
+runOne proof txIx (Trip beforeEnv afterEnv tx) = do
+  (lenv, ledgerstate) <- getSTSLedgerEnv proof txIx beforeEnv
+  expectedLedgerState <- runTarget afterEnv (ledgerStateT proof)
+  case applySTSByProof proof (TRC (lenv, ledgerstate, tx)) of
+    Right ledgerState' -> pure (ledgerState' === expectedLedgerState)
+    Left errs ->
+      pure
+        ( counterexample
+            ("\napplySTS fails\n" ++ unlines (map show errs) ++ "\nTx =\n" ++ (show (pcTx proof tx)))
+            ( whenFail
+                (putStrLn ("\napplySTS fails\n" ++ unlines (map show errs)) >> (goRepl proof afterEnv ""))
+                (True === False)
+            )
+        )
+
```

Comment:
OkI made this function
`-- | Create a Property by testing that applying the STS "LEDGER" rule, succeeds and
--   returns the expected LedgerState. If it fails, print out the failures and
--   drop into the Repl, so that users can explore the inputs.
ledgerStateEqProp
  ::  Proof era
     -> Env era
     -> State (EraRule "LEDGER" era)
     -> Environment (EraRule "LEDGER" era)
     -> State (EraRule "LEDGER" era)
     -> Signal (EraRule "LEDGER" era)
     -> Property
`